src/sdsc/spellcheck.py

```diff
@@ -0,0 +1,232 @@
+    currenttokeninparagraph = -1
+
+    messages = []
+    for sentence in sentences:
+        words = tokenizer(sentence)
+        totalwords = len(words)
+
+        skipcount = 0
+        for wordposition, word in enumerate(words):
+            # If we hit a placeholder, e.g. ##@key-1##, the number
+            # (here: 1) signifies the number of tokens this placeholder
+            # replaces. If no placeholder found, tagtokens is 1.
+            istag, _, tagtokens = findtagreplacement(word)
+            currenttokeninparagraph += tagtokens
+            if istag:
+                continue
+
+            word = removepunctuation(word, start=True, end=True)
+
+            # FIXME: English hardcoded
+            word = re.sub('(^[^-_.\d\w]+|(\'s|[®*!?.:;^])+$)', '', word)
+
+            # removepunctuation can lead to empty strings...
+            if not word:
+                continue
+            # filter for numbers, punctuation, URLs, file names
+            if SPELLFILTER.match(word):
+                continue
+
+            correct = spelldict.check(word)
+
+            if not correct:
+                # FIXME: can we order suggestions, e.g. for "ipc" (which is in
+                # our dictionary as "IPC"), we only get irrelevant lower-case
+                # suggestions and "IPC" removed later when we limit the choice
+                # to 5 suggestions.
+                suggestions = spelldict.suggest(word)
+
+                line = linenumber(context)
+                contenthighlighted = highlightREMOVEME(xmlescape(contentpretty), currenttokeninparagraph, currenttokeninparagraph)
+                messages.append(spellcheckmessage(
+                    suggestions, word, line,
+                    contenthighlighted, contextid, basefile,
+                    messagetype))
+
+    return messages
+
+
+def spellcheckmessage(suggestions, word, line, content,
+                        contextid, basefile, messagetype):
+    """ Create a message after a spell check has found a non-dictionary word.
+
+    :param list suggestions: list of strings with suggestions
+    :param str word: spelling, as actually used
+    :param int line: line number (generally wrong at the moment)
+    :param str content: paragraph of text that "word" appears in
+    :param str contextid: value of id attribute on next node
+    :param str basefile: name of file that "word" appears in
+    :param str messagetype: print 'error', 'warning', or 'info' message?
+    """
+
+    # FIXME: shorten content string (in the right place), to get closer toward
+    # more focused results
+    message = None
+    filename = ""
+    if basefile:
+        filename = "<file>%s</file>" % str(basefile)
+
+    withinid = ""
+    if contextid:
+        withinid = "<withinid>%s</withinid>" % str(contextid)
+
+    message = etree.XML("""<result type="%s">
+            <location>%s%s<line>%s</line></location>
+        </result>""" % (messagetype, filename, withinid, str(line)))
```

Comment:
You don't need `str(line)` as it is automatically converted to a string anyway (through `%s`). You can write `line`.